text/0000-router-service.md

```diff
@@ -0,0 +1,434 @@
+
+Both events receive a single `transition` argument as described in the "Transition Object" section below, which explains the meaning of `from` and `to` in more detail.
+
+Redirection example:
+
+ 1. current route is A
+ 2. user initiates a transition to B
+ 3. routeWillChange fires `from` A `to` B.
+ 4. B redirects to C
+ 5. routeWillChange fires `from` A `to` C.
+ 6. routeDidChange fires `from` A `to` C.
+
+Abort example:
+
+ 1. current route is A
+ 2. user initiates a transition to B
+ 3. routeWillChange fires `from` A `to` B.
+ 4. in response to the previous routeWillChange event, the transition is aborted.
+ 5. routeWillChange fires `from` A `to` A.
+ 8. routeDidChange fires `from` A `to` A.
+
+Error example:
+
+ 1. current route is A
+ 2. user initiates a transition to B.index
+ 3. routeWillChange fires `from` A `to` B.
+ 4. B throws an exception, and the router discovers a "B-error" template.
+ 5. routeWillChange fires `from` A `to` B-error
+ 6. routeDidChange fires `from` A `to` B-error
+
+These are events, not extension hooks -- now that we are exposing a Service, it makes more sense to subscribe to its events than extend it.
+
+### New Properties
+
+`currentRoute`: an observable property. It is guaranteed to change whenever a route transition happens (even when that transition only changes parameters and doesn't change the active route). You should consider its value deeply immutable -- we will replace the whole structure whenever it changes. The value of `currentRoute` is a `RouteInfo` representing the current leaf route. `RouteInfo` is described below.
+
+`currentRouteName`:  a convenient alias for `currentRoute.name`.
+
+`currentURL`: provides the serialized string representing `currentRoute`.
+
+### Query Parameter Semantics
+
+Today, `queryParams` impose unnecessarily high cost because we cannot generate URLs or determine if a link is active without taking into account the default values of query parameters. Determining their default values is expensive, because it involves instantiating the corresponding controller, even in cases where we will never visit its route.
+
+Therefore, the `queryParams` argument to the new `urlFor`, `transitionTo`, `replaceWith`, and `isActive` methods defined in this document will behave differently.
+
+ - default values will not be stripped from generated URLs. For example, `urlFor('my-route', { sortBy: 'title' })` will always include `?sortBy=title`, whether or not `title` is the default value of `sortBy`.
+
+ - to explicitly unset a query parameter, you can pass the symbol `Ember.DEFAULT_VALUE` as its value. For example, `transitionTo('my-route', { sortBy: Ember.DEFAULT_VALUE })` will result in a URL that does not contain any `?sortBy=`.
```

Comment:
They're additive in that they're new but considering our current state has severe issues and I want to ensure that we feel good about our future direction.

For the basic scenario and only-keyed-objects I think we can get away with `DELETE_VALUE`. To support arrays it almost has to become a Lisp where you specify the transforms. To do that gracefully I feel like we would need to adopt:
- `push`, `pop`
- `shift`, `unshift`
- `slice`
- `splice`
- ...

And that's beyond what I would want to support as it would come with a parser and all sorts of other annoying problems. Throwing away that proposal, I think it's bad.

---

We could always stamp something as "pre-merged" and expose an API to get the `unsupplied` values. It'll make structured query param merging palatable for advanced users, doesn't impact this simple scenario, doesn't add a tremendous amount of complexity and another Lisp-y microsyntax.

For bonus points I don't think that needs to be specified at this time, but seems like a direction where this doesn't paint us into any corners.

/FIN